### DISCUSSION TITLE
Integrating an external library into the global state

### DISCUSSION THREAD

**Discussion by afidegnum** [framesurge/perseus#175] [Q&A]
I'm trying to integrate deadpool_postgres so i can ran a database query but I face few issues, 
the following errors appeared. `the trait bound __deadpool_postgres::COnfig: global_state::_::_serde::Deserialize<'_> is not satisfied. The following other types implement trait "global_state::_::_serde::Deserialize<'de>"`

What do you suggest?
```rs
use config::ConfigError;
use deadpool_postgres::{Manager, Pool, Runtime};
use dotenv::dotenv;
use perseus::{state::GlobalStateCreator, RenderFnResult};
use serde::Deserialize;

pub fn get_global_state_creator() -> GlobalStateCreator {
    GlobalStateCreator::new().build_state_fn(get_build_state)
}

#[perseus::make_rx(AppStateRx)]
pub struct AppState {
    pub test: String,
}

#[derive(Deserialize, Clone, Debug)]
pub struct Config {
    pub pg: deadpool_postgres::Config,
}

impl Config {
    pub fn from_env() -> Result<Self, ConfigError> {
        config::Config::builder()
            .add_source(config::Environment::default())
            .build()
            .unwrap()
            .try_deserialize()
    }
}

#[perseus::global_build_state]
pub async fn get_build_state() -> RenderFnResult<AppState> {
    Ok(AppState {
        test: "Hello World!".to_string(),
    })
}

```

**Comment by afidegnum** [comment]
I noticed I asked that question before. 
adding `#[cfg(not(target_arch = "wasm32"))]` also didn't help,

**Comment by arctic-hen7** [comment]
Your issue is that you're trying to store a type, `Config`, in a `struct` that needs to implement `Deserialize`, but that type can't be converted to and from a string. Perseus needs to be able to generate state on the server-side and then send it over the internet to a client's browser, which means everything has to be able to be converted to a string. Unfortunately, what you're trying to do (i.e. storing a type that can't be managed as a string in state) is impossible.

However, I also think you don't need to do this. What you seem to be trying to do is send a database pool to the client, which is probably unnecessary. What I would do in this case is keep the database pool in a separate API handler on the server side, which the client then connects to. Let's say you have a database query that adds a user, just for learning purposes. What you're trying to do is send a database pool to the browser so that you can execute that query directly from the user's browser. But that would mean indirectly exposing your database's access keys to the user's browser, so anyone could extract them and your security would be gone. Instead, you could send a request to some API endpoint, maybe `/api/addUser`, providing the necessary information as JSON, and then the server could perform the database operation. This is what web developers would usually do in cases like these to prevent leaking database keys to users. Remember that anything you send to the user's browser, you have no control over --- it can be extracted.

Hope this helps! Let me know if you need any further clarification.

**Reply by afidegnum** [reply]
ok, I understand, but I feel there will be a little performance degradation when roundtrip requesting data from an API, what do you think? what can be done to improve on the performance? 

specially being on the same hardware, can we use sockets instead over REST ?

**Reply by arctic-hen7** [reply]
Yes, it would be a tiny bit slower, but that's better than giving your users the power to control your database.

I wouldn't recommend WebSockets for this, REST or similar would work well. You could use an in-memory database like Redis if you're very concerned about performance, or you could just make sure that the database and the server are running on the same hardware. Then you shouldn't have any problems.
